cmd/juju/config/config.go

```diff
@@ -0,0 +1,196 @@
+// This only parses the actual key/value arguments in the command. Some child
+// commands will also need to specify what is being configured (e.g. the app
+// name in the case of `juju config`). In this case, the child should define
+// its own Init command, where it strips the required arguments and passes the
+// rest to the parent.
+func (c *ConfigCommandBase) Init(args []string) error {
+	// Check if --file has been specified
+	if c.ConfigFile.Path != "" {
+		c.Actions = append(c.Actions, SetFile)
+	}
+
+	// Check if --reset has been specified
+	if len(c.reset) != 0 {
+		c.Actions = append(c.Actions, Reset)
+		err := c.parseResetKeys()
+		if err != nil {
+			return errors.Trace(err)
+		}
+	}
+
+	// The remaining arguments are divided into setKeys (args containing `=`)
+	// and getKeys, so we can work out what the user is trying to do.
+	var setKeys, getKeys []string
+	for _, arg := range args {
+		splitArg := strings.Split(arg, "=")
+		if len(splitArg) > 1 {
+			setKeys = append(setKeys, splitArg[0])
+		} else {
+			getKeys = append(getKeys, arg)
+		}
+	}
+
+	if len(setKeys) == 0 && len(getKeys) == 0 && len(c.Actions) == 0 {
+		// Nothing specified - get full config
+		c.Actions = append(c.Actions, GetAll)
+
+	} else if len(setKeys) == 0 && len(getKeys) == 1 {
+		// Get the single key specified
+		c.Actions = append(c.Actions, GetOne)
+		c.KeyToGet = args[0]
+
+	} else if len(setKeys) == 0 && len(getKeys) > 1 {
+		// Trying to get multiple keys - error
+		return errors.New("cannot get multiple keys at once")
+
+	} else if len(setKeys) > 0 && len(getKeys) == 0 {
+		// Set specified keys to given values
+		c.Actions = append(c.Actions, Set)
+		var err error
+		c.ValsToSet, err = keyvalues.Parse(args, true)
+		if err != nil {
+			return err
+		}
+
+	} else if len(setKeys) > 0 && len(getKeys) > 0 {
+		// Looks like user is trying to get & set simultaneously
+		return errors.Errorf(
+			"cannot get key(s) %s and set key(s) %s simultaneously",
+			printQuotedList(getKeys), printQuotedList(setKeys),
+		)
+	}
+
+	// Check that we haven't tried to set/get as well as reset/set from file.
+	return c.checkSingleAction()
+}
+
+// parseResetKeys splits the keys provided to --reset after trimming any
+// leading or trailing comma. It then verifies that we haven't incorrectly
+// received any key=value pairs.
+func (c *ConfigCommandBase) parseResetKeys() error {
+	for _, value := range c.reset {
+		keys := strings.Split(strings.Trim(value, ","), ",")
+		c.KeysToReset = append(c.KeysToReset, keys...)
+	}
+
+	for _, k := range c.KeysToReset {
+		if sliceContains(c.CantReset, k) {
+			return errors.Errorf("%q cannot be reset", k)
+		}
+		if strings.Contains(k, "=") {
+			return errors.Errorf(
+				`--reset accepts a comma delimited set of keys "a,b,c", received: %q`, k)
+		}
+	}
+	return nil
+}
+
+// sliceContains is a utility method to check if the given (string) slice
+// contains a given value.
+func sliceContains(slice []string, val string) bool {
```

Comment:
Ah, 'twill be nice when (hopefully in Go 1.19) we can use the generic [`slices.Contains`](https://pkg.go.dev/golang.org/x/exp/slices#Contains).

Reply:
Can't wait, long overdue